curvefs/src/space/space_service.cpp

```diff
@@ -0,0 +1,171 @@
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "curvefs/src/space/space_service.h"
+
+#include <brpc/closure_guard.h>
+#include <brpc/controller.h>
+
+#include "curvefs/src/space/utils.h"
+
+namespace curvefs {
+namespace space {
+
+SpaceAllocServiceImpl::SpaceAllocServiceImpl(SpaceManager* space)
+    : space_(space) {}
+
+void SpaceAllocServiceImpl::InitSpace(
+    ::google::protobuf::RpcController* controller,
+    const ::curvefs::space::InitSpaceRequest* request,
+    ::curvefs::space::InitSpaceResponse* response,
+    ::google::protobuf::Closure* done) {
+    brpc::ClosureGuard guard(done);
+    brpc::Controller* cntl = static_cast<brpc::Controller*>(controller);
+
+    auto status = space_->InitSpace(request->fsinfo());
+
+    response->set_status(status);
+    if (status != SPACE_OK) {
+        LOG(ERROR) << "InitSpace failure, logid: " << cntl->log_id()
+                   << ", request: " << request->ShortDebugString()
+                   << ", error: " << SpaceStatusCode_Name(status);
+    } else {
+        LOG(INFO) << "InitSpace success, logid: " << cntl->log_id()
+                  << ", request: " << request->ShortDebugString();
+    }
+}
+
+void SpaceAllocServiceImpl::UnInitSpace(
+    ::google::protobuf::RpcController* controller,
+    const ::curvefs::space::UnInitSpaceRequest* request,
+    ::curvefs::space::UnInitSpaceResponse* response,
+    ::google::protobuf::Closure* done) {
+    brpc::ClosureGuard guard(done);
+    brpc::Controller* cntl = static_cast<brpc::Controller*>(controller);
+
+    auto status = space_->UnInitSpace(request->fsid(), request->fstype());
+    response->set_status(status);
+    if (status != SPACE_OK) {
+        LOG(ERROR) << "UnInitSpace failure, " << cntl->log_id()
+                   << ", request: " << request->ShortDebugString()
+                   << ", error: " << SpaceStatusCode_Name(status);
+    } else {
+        LOG(INFO) << "UnInitSpace success, " << cntl->log_id()
+                  << ", request: " << request->ShortDebugString();
+    }
+}
+
+void SpaceAllocServiceImpl::AllocateSpace(
+    ::google::protobuf::RpcController* controller,
+    const ::curvefs::space::AllocateSpaceRequest* request,
+    ::curvefs::space::AllocateSpaceResponse* response,
+    ::google::protobuf::Closure* done) {
+    brpc::ClosureGuard guard(done);
+
+    Extents exts;
+    SpaceAllocateHint hint;
+    if (request->has_allochint()) {
+        hint = ConvertToSpaceAllocateHint(request->allochint());
+    }
+
+    auto status =
+        space_->AllocateSpace(request->fsid(), request->size(), hint, &exts);
```

Comment:
Pass 'response->mutable_extents' to AllocateSpace to avoid copy.

Reply:
Some interfaces of protobuf are not easy to use, so I just defined `PExtent` and `Extents` and after allocation finished, convert `Extents` to protobuf.
And as you said, this will cause some memory copy, but I think is controllable, it's just a bunch of double uint64 value.